aptos-move/block-executor/src/scheduler_status.rs

```diff
@@ -0,0 +1,1344 @@
+            // Acquire write lock for (non-monitor) shortcut modifications.
+            let inner_status = self.inner_status.lock();
+
+            let dependency_flag =
+                DependencyFlag::from_u8(self.dependency_shortcut.load(Ordering::Relaxed))?;
+
+            match (inner_status.pending_scheduling(), dependency_flag) {
+                (Some(0), DependencyFlag::Defer) => {
+                    // Adding a stall requires being recorded in aborted depedencies in scheduler_v2,
+                    // which in turn only happens in the scheduler after a successful abort (that must
+                    // increment the incarnation of the status).
+                    return Err(code_invariant_error("0-th incarnation in add_stall"));
+                },
+                (Some(_), DependencyFlag::Defer) => {
+                    self.execution_queue_manager
+                        .remove_from_schedule(self.txn_idx);
+                    // Shortcut not affected.
+                },
+                (Some(_), DependencyFlag::Safe | DependencyFlag::Executing) => {
+                    return Err(code_invariant_error(
+                        "Inconsistent status and dependency shortcut in add_stall",
+                    ));
+                },
+                (None, DependencyFlag::Safe) => {
+                    // May not update SAFE flag at a future incorrect time (i.e. ABA), as observing
+                    // num_stalls = 0 under status is required to set SAFE flag, but impossible
+                    // until the corresponding remove_stall (that starts only after add_stall finishes).
+                    self.dependency_shortcut
+                        .store(DependencyFlag::Defer as u8, Ordering::Relaxed);
+                },
+                (None, DependencyFlag::Executing | DependencyFlag::Defer) => {
+                    // Executing or aborted: shortcut not affected.
+                },
+            }
+
+            return Ok(true);
+        }
+        Ok(false)
+    }
+
+    /// Removes a stall from the transaction, potentially allowing its re-execution.
+    ///
+    /// This is the counterpart to add_stall. When the number of stalls drops to zero,
+    /// the transaction becomes "unstalled" and may be rescheduled for execution if
+    /// it's in the PendingScheduling state.
+    ///
+    /// # Returns
+    /// - `Ok(true)` if this call changed the state from stalled to unstalled (num_stalls 1→0)
+    /// - `Ok(false)` if the transaction remains stalled after this call
+    /// - `Err` if there was an error removing the stall (e.g., no matching add_stall)
+    pub(crate) fn remove_stall(&self) -> Result<bool, PanicError> {
+        let prev_num_stalls = self.num_stalls.fetch_sub(1, Ordering::SeqCst);
+
+        if prev_num_stalls == 0 {
+            return Err(code_invariant_error(
+                "remove_stall called when num_stalls == 0",
+            ));
+        }
+
+        if prev_num_stalls == 1 {
```

Comment:
Question: can it be that you check this here, but then someone calls `add_stall`? I see there is a check below, but can't wrap my head around ordering if making it relaxed may still see old value with 0? Does the lock acquisition before ensure that after this relaxed is ok because no other thread can add a stall at this point?

Reply:
For the re-check that happens after acquiring the lock, you'd have the guarantee that it'd observe any update to the (num_stalls) variable that'd happen before the last lock release (including while holding the lock), which is exactly the guarantee for Relaxed semantics.